common/lanelet2_extension/lib/CarmaUSTrafficRules.cpp

```diff
@@ -0,0 +1,297 @@
+ */
+bool canTravelInDir(const lanelet::ConstLanelet& ll, const std::string& participant)
+{
+  if (!ll.inverted())
+  {
+    return true;  // If this lanelet is not inverted then one-way or bi-directional are both fine
+  }
+
+  auto regs = ll.regulatoryElementsAs<DirectionOfTravel>();
+
+  if (regs.size() == 0)
+  {
+    return false;  // Default to one way. So if there is no regulation then return false (only allow oneway)
+  }
+  else if (regs.size() == 1)
+  {
+    // Add logic
+    return !(regs[0]->isOneWay()) &&
+           regs[0]->appliesTo(participant);  // If there is a regulation return true if bi-directional
+  }
+  else
+  {
+    throw std::invalid_argument("CarmaUSTrafficRules could not determine traffic rules as two DirectionOfTravel "
+                                "regulations were applied on lanelet: " +
+                                std::to_string(ll.id()));
+  }
+}
+
+bool CarmaUSTrafficRules::canAccessRegion(const ConstLaneletOrArea& region) const
+{
+  auto accessRestrictions = region.regulatoryElementsAs<RegionAccessRule>();
+  for (auto access_rule : accessRestrictions)
+  {
+    if (!access_rule->accessable(participant()))
+    {
+      return false;
+    }
+  }
+  return true;  // TODO should we support attributes as fallback?
+}
+
+bool CarmaUSTrafficRules::canPass(const ConstLanelet& lanelet) const
+{
+  if (!canTravelInDir(lanelet, participant()))
+  {  // Check direction of travel
+    return false;
+  }
+  // Check access by participant
+  ConstLaneletOrArea region(lanelet);
+  return canAccessRegion(region);
+}
+
+bool CarmaUSTrafficRules::canPass(const ConstArea& area) const
+{
+  ConstLaneletOrArea region(area);
+  return canAccessRegion(region);
+}
+
+// NOTE: Based on the GenericTrafficRules object this function is for non-lanechange passing
+bool CarmaUSTrafficRules::canPass(const ConstLanelet& from, const ConstLanelet& to) const
+{
+  return geometry::follows(from, to) && canPass(from) && canPass(to);
+}
+
+bool CarmaUSTrafficRules::boundPassable(const ConstLineString3d& bound,
+                                        const std::vector<PassingControlLineConstPtr>& controlLines,
+                                        bool fromLeft) const
+{
+  bool result = PassingControlLine::boundPassable(bound, controlLines, fromLeft, participant());
+  // TODO throw exception or revert to attributes or return true????
+  // We could fallback on generic traffic rules here
+  return result;
+}
+
+bool CarmaUSTrafficRules::canPass(const ConstLanelet& from, const ConstArea& to) const
+{
+  if (!canPass(from) || !canPass(to))
+  {
+    return false;
+  }
+
+  if (geometry::leftOf(from, to))
+  {
+    return boundPassable(from.rightBound(), from.regulatoryElementsAs<PassingControlLine>(), true);
+  }
+  if (geometry::rightOf(from, to))
+  {
+    return boundPassable(from.leftBound(), from.regulatoryElementsAs<PassingControlLine>(), false);
+  }
+
+  auto line = determineCommonLine(from, to);
+  if (!!line)
```

Comment:
Is that the same as if(line)?

Reply:
Yes, the !! is just a way of converting to bool.